hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/commit/AbfsManifestStoreOperations.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.commit;
+
+import java.io.IOException;
+import java.time.Duration;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathIOException;
+import org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem;
+import org.apache.hadoop.mapreduce.lib.output.committer.manifest.files.FileEntry;
+import org.apache.hadoop.mapreduce.lib.output.committer.manifest.impl.ManifestStoreOperationsThroughFileSystem;
+
+/**
+ * Extension of StoreOperationsThroughFileSystem with ABFS awareness.
+ * Purely for use by jobs committing work through the manifest committer.
+ * The {@link AzureManifestCommitterFactory} will configure the
+ * configure to use this as the binding to the FS.
+ *
+ * ADLS Gen2 stores support etag-recovery on renames, but not WASB
+ * stores.
+ */
+@InterfaceAudience.LimitedPrivate("mapreduce")
+@InterfaceStability.Unstable
+public class AbfsManifestStoreOperations extends
```

Comment:
On avoiding the new dependency on hadoop-mapreduce-client (or the reverse where hadoop-mapreduce-client ends up depending on hadoop-cloud-storage / azure specifically)

Does it make sense to make ResilientCommitByRename a little more generic, and move it to hadoop-common as a LimitedPrivate("cloud-storage"?)/Unstable.
AzureBlobFileSystem is the only one to implement the interface right now.

AbfsManifestStoreOperations - can then make decisions based on whether the FileSystem implements the interface (and can potentially be merged into ManifestStoreOperationsThroughFileSystem itself) / or gets renamed to something like ResilientRenameCapableManifestStoreOperationsThroughFileSystem.

All depends on moving ResilientCommitByRename to hadoop-common. I think this is quite a bit cleaner than adding the dependencies in either direction.

Reply:
we did look at that at first but it'd force us to do a lot more work in a public API, specifically "get rename right". I've played with this in the past #2735 , with the API we could do for async and slow operations:
https://github.com/apache/hadoop/pull/2735#issuecomment-892760289

problem here is that rename() is the nightmare, even working out what is good is hard, everything is inconsistent etc etc. Doing a special binding api in the abfs module lets us avoid worrying about this, lets us tune other operations if need to worry about problems like rate limiting there too, and lets us remove it when abfs makes this failure mode go away.

I did put the etag source API into hadoop common, with a PathCapabilities probe to verify that a store supports etag preservation over rename. So there is enough in hadoop common for hive to implement its own recovery. because that API is in common, it means that there's no dependency on the mapred- libraries in normal use of the client.
Added a statement in the javadocs of `org.apache.hadoop.fs.azurebfs.commit.ResilientCommitByRename` to make clear that there mustn't be any references in there